skimage/metrics/_set_metrics.pyx

```diff
@@ -0,0 +1,52 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+cimport numpy as cnp
+from libc.math cimport sqrt
+from libc.float cimport DBL_MAX
+
+import numpy as np
+
+def hausdorff_distance_onesided(cnp.float64_t[:, ::1] points_sup,
```

Comment:
One of my pet comments here: in terms of speed, how would this Cython implementation compare to using a CKDTree, in particular https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html#scipy.spatial.cKDTree.query (the `query` method for one nearest-neighbor, which directly returns the distance)?

My guestimate is that for a small number of points the Cython will be faster, but that for a large number of points the KDTree might win. Are you interested in testing this @clementkng ? If you don't have the time I might give it a try. 

I guess it also depends on the application for which we expect this function to be used. Will it be rather for very sparse images or for more busy ones?

Reply:
Hi @emmanuelle, I can try to compare the cKDTree vs cython, but I'm not super familiar w/ how we'd test this. I'm guessing this is similar to the benchmarking process for the initial implementation of Hausdorff distance?

I expect the function to be used for reasonably busy images, as one of the applications is to find and recognize image shapes within a larger image.

Reply:
I would just take a realistic example and see how the two methods compare, and also rescale the image of interest with several factors so that you can see if the fastest method depends on the number of non-zero pixels. Automatic benchmarks with asv are useful but here it's just to decide which method is better.

Reply:
@emmanuelle to ensure reproducibility, should I push the code I use to compare the methods into the benchmark_metrics.py as well?

Reply:
not at this stage because if one method is a clear winner we will not bother with the other. What I would do is write a python script calling the function on images of different sizes, and run this script in two different branches (each branch with a different implementation), and paste the script and its result here. But really whatever workflow works for you is fine :-).

Reply:
A workflow I have employed successfully in the past uses notebooks to compare two implementations side by side using the "%timeit" magic command. E.g. see this [old gist](https://gist.github.com/lagru/4cd7ab94a101bbea15af216a3d44d1ca) were I imported the new implementation from the PR branch and declared the old implementation inside the notebook to compare.

Reply:
I would love to see a comparison of your implementation to https://github.com/deepmind/surface-distance/blob/master/surface_distance/metrics.py#L209 in terms of speed on a pair of 3D images (e.g. 400x400x400).

Reply:
@soupault I'm not sure I've seen the surface-distance library used in an Jupyter notebook. I'm having a little trouble including it in the comparisons, is there a code sample somewhere of how to import the package and use it if possible?

Reply:
Are you asking how to install this library? You should be able to import `compute_robust_hausdorff from surface_distance` after following the install instructions in the [readme](https://github.com/deepmind/surface-distance#installation).

Reply:
So something like `from surface_distance import compute_robust_hausdorff` should work once I've installed the package right? I believe I'm having some issues since surface distance is not [Python3 compatible](https://github.com/deepmind/surface-distance/pull/1), but I need Python3 to run/build scikit-image.

Reply:
Here is a [link](https://gist.github.com/clementkng/cf0321e2b62daed21189867187e36751) to a first draft of benchmarks against cKD tree.

Reply:
@clementkng From the issue you linked it looks like only that one import must be updated to be compatible with Python 3? You might succeed by applying that change to your local copy. 

Thanks for the notebook! I've left a suggestion to make the comparison more meaningful.